repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java

```diff
@@ -686,6 +687,113 @@ private void traverseImpactedVertices(final AtlasVertex entityVertexStart, final
         RequestContext.get().endMetricRecord(metricRecorder);
     }
 
+    private void traverseImpactedVerticesByLevel(final AtlasVertex entityVertexStart, final String relationshipGuidToExclude,
+                                          final String classificationId, final List<String> result, List<String> edgeLabelsToExclude) {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("traverseImpactedVerticesByLevel");
+        Set<String>                 visitedVerticesIds        = new HashSet<>();
+        Set<String>                 verticesAtCurrentLevel    = new HashSet<>();
+        Set<String>                 traversedVerticesIds      = new HashSet<>();
+        RequestContext              requestContext            = RequestContext.get();
+
+        //Add Source vertex to level 1
+        if (entityVertexStart != null) {
+            verticesAtCurrentLevel.add(entityVertexStart.getIdForDisplay());
+        }
+        // Start Processing the level
+        while (!verticesAtCurrentLevel.isEmpty()) {
+            Set<String> verticesToVisitNextLevel = new HashSet<>();
+            // Chunk the Levels into n parts, with constant bucket size
+            Iterable<List<String>> chunkedLevelVerticesIterable = Iterables.partition(verticesAtCurrentLevel, AtlasConfiguration.GRAPH_TRAVERSE_LEVEL_BUCKET_SIZE.getInt());
+            chunkedLevelVerticesIterable.forEach(x -> {
+                // Process the chunks in Parallel get the Vertices for next level
+                x.parallelStream().forEach(y -> {
+                    AtlasVertex entityVertex = graph.getVertex(y);
+                    if (!visitedVerticesIds.contains(y)) {
+                        //get adjacent vertices of the current level vertex to be added to next level
+                       Set<String> adjacentVerticesIds =  getAdjacentVertices(entityVertex, classificationId,
+                                relationshipGuidToExclude, edgeLabelsToExclude, visitedVerticesIds);
+                        verticesToVisitNextLevel.addAll(adjacentVerticesIds);
+                        traversedVerticesIds.addAll(adjacentVerticesIds);
+                        visitedVerticesIds.add(entityVertex.getIdForDisplay());
+                    }
+                });
```

Comment:
this implementation is not optimal actually because the parallelisation happens in the inner loop, the outer loop is sequential so for bucket size of 100, all 100 must complete first before next 100 is executed. The optimal implementation is for 100 traversal to always be running, when one traversal is completed the next traversal is added

not a blocker for this PR but you can think about it